modules/dnn/src/cuda4dnn/csl/tensor_ops.hpp

```diff
@@ -0,0 +1,470 @@
+                get_effective_rank(B) <= 2 &&
+                get_effective_rank(result) <= 2);
+
+            /* check dimension requirements for matrix multiplication */
+            if (!transa && !transb) {
+                CV_Assert(A.get_axis_size(-2) == result.get_axis_size(-2));
+                CV_Assert(A.get_axis_size(-1) == B.get_axis_size(-2));
+                CV_Assert(B.get_axis_size(-1) == result.get_axis_size(-1));
+            } else if (!transa && transb) {
+                CV_Assert(A.get_axis_size(-2) == result.get_axis_size(-2));
+                CV_Assert(A.get_axis_size(-1) == B.get_axis_size(-1));
+                CV_Assert(B.get_axis_size(-2) == result.get_axis_size(-1));
+            } else if (transa && !transb) {
+                CV_Assert(A.get_axis_size(-1) == result.get_axis_size(-2));
+                CV_Assert(A.get_axis_size(-2) == B.get_axis_size(-2));
+                CV_Assert(B.get_axis_size(-1) == result.get_axis_size(-1));
+            } else {
+                CV_Assert(A.get_axis_size(-1) == result.get_axis_size(-2));
+                CV_Assert(A.get_axis_size(-2) == B.get_axis_size(-1));
+                CV_Assert(B.get_axis_size(-2) == result.get_axis_size(-1));
+            }
+
+            const auto result_nr = result.get_axis_size(-2);
+            const auto result_nc = result.get_axis_size(-1);
+            const auto common_dim = A.get_axis_size(transa ? -2 : -1);
+            const auto A_nc = A.get_axis_size(-1);
+            const auto B_nc = B.get_axis_size(-1);
+
+            /* tensors are stored in row-major but cublas::gemm operates on column-major matrices
+             * a row-major matrix when read as column-major matrix gives the transpose of the intended matrix
+             *
+             * Required: C = AB
+             * what cuBLAS sees: C^T = A^TB^T = (BA)^T
+             *
+             * By reversing operands, we effectively perform:
+             * C^T = B^TA^T = (AB)^T
+             *
+             * which gives C = AB
+             */
+            cublas::gemm<T>(handle,
+                transb, transa,
+                result_nc, result_nr, common_dim,
+                alpha, B.get(), B_nc,
+                A.get(), A_nc,
+                beta, result.get(), result_nc);
+        }
+
+        /** @brief performs element-wise addition with broadcasting
+         *
+         * Pre-conditions:
+         * - \p A and \p C must be compatible tensors
+         *
+         * Exception Gaurantee: Basic
+         */
+        template <class T> inline
+        void add(const cudnn::Handle& handle, T beta, TensorSpan<T> C, T alpha, TensorView<T> A) {
+            CV_Assert(is_shape_compatible(A, C));
+
+            using cudnn::TensorDescriptor;
+            auto aDesc = TensorDescriptor<T>(A.shape());
+            auto cDesc = TensorDescriptor<T>(C.shape());
+            cudnn::add(handle, alpha, aDesc, A.get(), beta, cDesc, C.get());
```

Comment:
The `cudnnAddTensor` is much slower than the biasN kernels. Almost 2-3x slower when compared with the vectorized biasN kernels.

Reply:
I’ve found that a number of the really simple things in cudnn such as this are slow. Writing my own simple kernel to do this stuff ended up being faster. Sometimes surprisingly so.